[docs] onchain decryption (#288)
MystenLabs/seal: docs/UsingSeal.md

You can use the TypeScript SDK to build a transaction that calls Seal’s on-chain decryption functions. 

Before encryption, the content creator may create a list of `PublicKey` Move objects containing the key server object ID and its public key. Clients can verify them if needed. 

```typescript
const publicKey = tx.moveCall({
  target: `${SEAL_PACKAGE_ID}::bf_hmac_encryption::new_public_key`,
  arguments: [
    tx.pure.address(keyserverId),

>     tx.pure.vector("u8", Array.from(publicKeyBytes)) // stored in the desired key server object's dynamic object field `pk`. 

Reviewer: ```suggestion
    tx.pure.vector("u8", Array.from(publicKeyBytes)) // store this in the target key server object's dynamic field `pk`.
```

  ],
});
const allPublicKeys = [publicKey]; // can be many
```

> Now lient can first verify the keys, then decrypt using the Move functions. 

Reviewer: ```suggestion
First, client can verify the public keys. Then decrypt with the Move functions.
```

> Client should have:

Reviewer: ```suggestion
The client must have:
```

- `encryptedBytes`: the BCS-serialized encrypted object.
- `txBytes`: a valid transaction block that satisfies the `seal_approve*` policy function (see [Decryption](#decryption)).

> - `sessionKey`: a session key for the given package (see [Decryption](#decryption)).

Reviewer: ```suggestion
- `sessionKey`: a session key for the target package (see [Decryption](#decryption)).
```

> - `allPublicKeys`: an array of Move objects for all public keys in the encryption.

Reviewer: ```suggestion
- `allPublicKeys`: an array of Move objects for all public keys used in the encryption.
```

Author: this is get from getPublicKeys

> - `correspondingPublicKeys`: the public keys that correspond to the derived keys. // whats this difference btw allPublicKeys vs correspondingPublicKeys????

Reviewer: ```suggestion
- `correspondingPublicKeys`: the public keys that correspond to the derived keys.
```

Author: this the subset of the public keys for decryption, just need to meet the threshold

```typescript
// Parse BCS serialized encrypted object from bytes.
const encryptedObject = EncryptedObject.parse(encryptedBytes);

// Alternatively, parse encrypted object as Move object with `parse_encrypted_object`.
// For some applications, this object should be stored on-chain (e.g., encrypted votes should be stored before the deadline, and decrypted after).
const tx = new Transaction();
const encryptedObject = tx.moveCall({
  target: `${SEAL_PACKAGE_ID}::bf_hmac_encryption::parse_encrypted_object`,
  arguments: [tx.pure.vector("u8", Array.from(encryptedBytes))],
});

// Get derived keys from key servers.
const derivedKeys = await sealClient.getDerivedKeys({
  id: encryptedObject.id,
  txBytes,
  sessionKey,
  threshold: encryptedObject.threshold,
});

// Convert the derived keys to G1 elements.
const derivedKeysAsG1Elements = Array.from(derivedKeys).map(([derivedKey]) => 
  tx.moveCall({
    target: `0x2::bls12381::g1_from_bytes`,
    arguments: [tx.pure.vector("u8", fromHex(derivedKey.toString()))],
  }),
);

// Call to verify the derived keys. This can be cached if decryption for the same ID is done again. 

> // what does this mean `for the same ID is done again.`????

Reviewer: ```suggestion
```

Author: not necessarily the same encrypted object: but for the same package id + id

You can use the TypeScript SDK to build a transaction that calls Seal’s on-chain decryption functions. 

> Assume you have the following (See [Decryption](#decryption)): 

Reviewer: ```suggestion
Before you decrypt (see [Decryption](#decryption)), gather the following: 
```

> - `encryptedBytes`: the BCS-serialized encrypted object.

Reviewer: ```suggestion
- `encryptedBytes`: BCS-serialized encrypted object.
```

> - `txBytes`: a valid transaction block that satisfies the `seal_approve*` policy function.

Reviewer: ```suggestion
- `txBytes`: a valid transaction block that calls the relevant `seal_approve*` policy function.
```

> - `client`: the initialized `SealClient`. 

Reviewer: ```suggestion
- `client`: an initialized `SealClient`. 
```

> - `sessionKey`: the initialized `SessionKey`.

Reviewer: ```suggestion
- `sessionKey`: an initialized `SessionKey`.
```

- `SEAL_PACKAGE_ID`: the Seal package ID for the network. 

```typescript

> // 1. Parse this encrypted object.

Reviewer: ```suggestion
// 1. Parse the encrypted object.
```

const encryptedObject = EncryptedObject.parse(encryptedBytes);

> // 2. Get derived keys from key servers for this object ID. 

Reviewer: ```suggestion
// 2. Get derived keys from key servers for the encrypted object's ID. 
```

  const index = encryptedObject.services.findIndex(([s, _]) => s === objectId);
  return tx.moveCall({
    target: `${seal_package_id}::bf_hmac_encryption::new_public_key`,
    arguments: [
      tx.pure.address(objectId),
      tx.pure.vector("u8", publicKeys[index].toBytes())
    ],
  });
}).toArray();

// 4. Convert the derived keys to G1 elements.
const derivedKeysAsG1Elements = Array.from(derivedKeys).map(([_, value]) =>
tx.moveCall({
  target: `0x2::bls12381::g1_from_bytes`,
  arguments: [ tx.pure.vector("u8", fromHex(value.toString())) ],
})
);

// 5. Call the Move function verify_derived_keys. This should be cached if decryption for the same ID is performed again. 
const verifiedDerivedKeys = tx.moveCall({
  target: `${seal_package_id}::bf_hmac_encryption::verify_derived_keys`,
  arguments: [
  tx.makeMoveVec({ elements: derivedKeysAsG1Elements, type: '0x2::group_ops::Element<0x2::bls12381::G1>' }),
  tx.pure.address(encryptedObject.packageId),
  tx.pure.vector("u8", fromHex(encryptedObject.id)),
    tx.makeMoveVec({ elements: correspondingPublicKeys, type: `${SEAL_PACKAGE_ID}::bf_hmac_encryption::PublicKey` }),
  ],
});

> // 6. Construct the encrypted object.

Reviewer: We already have the BCS-serialized encrypted object from earlier which we're passing here. Should we rephrase this instruction to differentiate from that somehow?

  ],
});

// Call to decrypt. 
const result = tx.moveCall({
  target: `${SEAL_PACKAGE_ID}::bf_hmac_encryption::decrypt`,
  arguments: [
    parsedEncryptedObject,
    verifiedDerivedKeys,
    tx.makeMoveVec({ elements: allPublicKeys }),
  ],
});

> // result is in an option to be consumed if decryption is successful, none otherwise. 

Reviewer: ```suggestion
// result is in an option to be consumed if decryption is successful, `none` otherwise. 
```
